scripts-available/CDB_QuantileBins.sql

```diff
@@ -5,39 +5,15 @@
 --            bins based on the Quantile method.
 --
 -- @param breaks The number of bins you want to find.
---  
 --
-CREATE OR REPLACE FUNCTION CDB_QuantileBins ( in_array NUMERIC[], breaks INT) RETURNS NUMERIC[] as $$ 
-DECLARE 
-    element_count INT4; 
-    break_size numeric;
-    tmp_val numeric; 
-    i INT := 1; 
-    reply numeric[]; 
-BEGIN 
-    -- sort our values
-    SELECT array_agg(e) INTO in_array FROM (SELECT unnest(in_array) e ORDER BY e ASC) x;
-    -- get the total size of our data
-    element_count := array_length(in_array, 1); 
-    break_size :=  element_count::numeric / breaks;
-    -- slice our bread
-    LOOP  
-        IF i < breaks THEN
-            IF break_size * i % 1 > 0 THEN
-                SELECT e INTO tmp_val FROM ( SELECT unnest(in_array) e LIMIT 1 OFFSET ceil(break_size * i) - 1) x;
-            ELSE
-                SELECT avg(e) INTO tmp_val FROM ( SELECT unnest(in_array) e LIMIT 2 OFFSET ceil(break_size * i) - 1 ) x;
-            END IF;
-        ELSIF i = breaks THEN
-            -- select the last value
-            SELECT max(e) INTO tmp_val FROM ( SELECT unnest(in_array) e ) x;
-        ELSE
-            EXIT;
-        END IF;
-
-        reply = array_append(reply, tmp_val);
-        i := i+1;
-    END LOOP;
-    RETURN reply;
-END; 
-$$ language plpgsql IMMUTABLE STRICT PARALLEL SAFE;
+--
+CREATE OR REPLACE FUNCTION CDB_QuantileBins(in_array numeric[], breaks int)
+RETURNS numeric[]
+AS $$
+  SELECT array_agg(p) FROM (
+      SELECT percentile_disc(idx::numeric / breaks::numeric)
+        WITHIN GROUP (ORDER BY x ASC) AS p
+      FROM generate_series(1, breaks) AS idx, unnest(in_array) AS x
+      GROUP BY idx
+  ) AS quantiles;
+$$ language sql;
```

Comment:
What do you think of writing it using the array version of the `percentile_disc` functions as:

```sql
CREATE OR REPLACE FUNCTION CDB_QuantileBins(in_array numeric[], breaks int)
RETURNS numeric[]
AS $$
    SELECT percentile_disc(array(select generate_series(1,breaks)/breaks::numeric))
    WITHIN GROUP (ORDER BY x ASC) AS p
    FROM unnest(in_array) AS x
$$ language sql IMMUTABLE STRICT PARALLEL SAFE;
```

Reply:
@ethervoid It's added because some people do use PG >= 9.6, including us for testing. The Makefile will [remove it](https://github.com/CartoDB/cartodb-postgresql/blob/c1c55bf59f0814aacd8e114988da83a7f6ebe83f/Makefile#L126) if using PG 9.5 or before.

Reply:
Great, and what do you think about the IMMUTABLE flag?

Reply:
It's way more important to mark it as `IMMUTABLE` than to mark it as `PARALLEL SAFE`. Immutable means that the query output can be cached "forever" and reused when needed, which is better than executing it twice as fast :smile:

Reply:
Yes, I know. but my question was more focused on if with the new changes, and in your experience, its suitable to be immutable. (In my opinion it is and must be)

Reply:
Awesome catch about array argument @ethervoid !!

Reply:
Oh sorry, I misunderstood. Yes, the function is IMMUTABLE. What I usually do to check it is start as IMMUTABLE by default, look for odd things like queries making queries (like checking cache) and at last check if any function used isn't IMMUTABLE (`\df+ percentile_disc`); so essentially this function should be immutable.
Other way to look at this is to thing whether you can cache the result:
* IMMUTABLE: Forever. The output will "never" change. Example: A function that returns the population of a given city in a year in the past.
* STABLE: Can be cached inside a query. Example: The amount of bike lanes km in Madrid currently. For example, a function that returns the current length of a city bike lanes. It changes but you don't expect it to change (or care) in the duration of a query or plpgsql procedure.
* VOLATILE: Should never be cached, even inside a loop/procedure/etc. Example: A function that returns the amount of credits left for an user.

Reply:
:clap: a great guide on how to check if the function could be immutable or stable! thank you :)))))